FGEGraphics/UISystem/UIInputLabel.cs

```diff
@@ -0,0 +1,469 @@
+        public void SetPosition(int cursorPos) => CursorLeft = CursorRight = cursorPos;
+        
+        /// <summary>Clamps the cursor positionst to the <see cref="TextContent"/> bounds.</summary>
+        public void ClampPositions()
+        {
+            CursorLeft = Math.Clamp(CursorLeft, 0, TextContent.Length);
+            CursorRight = Math.Clamp(CursorRight, 0, TextContent.Length);
+        }
+
+        /// <summary>Sets the input text content.</summary>
+        /// <param name="content">The new text content.</param>
+        public void SetTextContent(string content)
+        {
+            if (TextContent == content)
+            {
+                return;
+            }
+            TextContent = content ?? string.Empty;
+            ClampPositions();
+        }
+
+        /// <summary>Updates the text components based on the cursor positions.</summary>
+        public readonly void UpdateTextComponents()
+        {
+            TextLeft.Content = TextContent[..IndexLeft];
+            TextBetween.Content = TextContent[IndexLeft..IndexRight];
+            TextRight.Content = TextContent[IndexRight..];
+        }
+
+        /// <summary>Calculates a screen cursor offset given the current <see cref="TextChain"/>.</summary>
+        // TODO: Account for formatting codes
+        public readonly Location GetCursorOffset()
+        {
+            double xOffset = 0;
+            int cursorIndex = IndexLeft;
+            int currentIndex = 0;
+            cursorIndex -= TextChain.Sum(piece => piece.SkippedIndices.Where(index => index <= cursorIndex).Count());
+            for (int i = 0; i < TextChain.Count; i++)
+            {
+                UIElementText.ChainPiece piece = TextChain[i];
+                for (int j = 0; j < piece.Text.Lines.Length; j++)
+                {
+                    RenderableTextPart[] parts = piece.Text.Lines[j].Parts;
+                    if (parts.Length == 0 && i == TextChain.Count - 1)
+                    {
+                        return new Location(0, piece.YOffset, 0);
+                    }
+                    foreach (RenderableTextPart part in parts)
+                    {
+                        if (currentIndex + part.Text.Length < cursorIndex)
+                        {
+                            currentIndex += part.Text.Length;
+                            xOffset += part.Width;
+                            continue;
+                        }
+                        int relIndex = cursorIndex - currentIndex;
+                        double x = xOffset + (part.Text.Length > 0 ? piece.Font.MeasureFancyText(part.Text[..relIndex]) : 0);
+                        double y = piece.YOffset + j * piece.Font.FontDefault.Height;
+                        return new Location(x, y, 0);
+                    }
+                    xOffset = 0;
+                }
+                currentIndex++;
+            }
+            return Location.NaN; // Should never happen.
```

Comment:
If this is expected to be impossible, `throw new UnreachableException();`

Reply:
Or if you want to use `NaN` as a general "just don't render a cursor" flag, need to document it and have checks for it